pkg/analysis/optionalfields/analyzer.go

```diff
@@ -183,7 +205,54 @@ func (a *analyzer) checkFieldPropertiesWithoutOmitEmpty(pass *analysis.Pass, fie
 		// Once it has the omitempty tag, it will also need to be a pointer in some cases.
 		// Now handle it as if it had the omitempty already.
 		// We already handle the omitempty tag above, so force the `hasOmitEmpty` to true.
-		a.checkFieldPropertiesWithOmitEmptyRequired(pass, field, fieldName, jsonTags, underlying, true, hasValidZeroValue, completeValidation, isPointer, isStruct)
+		a.checkFieldPropertiesWithOmitEmptyRequired(pass, field, fieldName, jsonTags, underlying, true, false, hasValidZeroValue, completeValidation, isPointer, isStruct)
+	}
+}
+
+func (a *analyzer) checkFieldPropertiesWithOmitZeroRequired(pass *analysis.Pass, field *ast.Field, fieldName string, jsonTags extractjsontags.FieldTagInfo, hasOmitZero, isPointer, isStruct, hasValidZeroValue bool) {
+	if !isStruct {
+		// ignore non-struct fields.
+		return
+	}
+
+	// In this case, we should always add the omitzero for non-pointer struct fields if it isn't present.
+	if !isPointer {
+		a.handleFieldShouldHaveOmitZero(pass, field, fieldName, hasOmitZero, jsonTags)
+	}
+
+	switch {
+	case hasValidZeroValue:
+		a.handleFieldShouldNotBePointer(pass, field, fieldName, isPointer, "field %s is optional and have a valid zero value. The field does not need to be a pointer.")
+	case !hasValidZeroValue:
+		a.handleFieldShouldNotBePointer(pass, field, fieldName, isPointer, "field %s is optional and does not have a valid zero value. The field does not need to be a pointer.")
+	}
```

Comment:
Bit confused here, It does not allow to set pointer on struct.

if struct has omitzero tag
   if zero value is not valid it need be pointer
   if zero value is valid then also it need not be pointer

Reply:
If the zero value is valid, then it does need to be a pointer, and have omitempty

There are two valid choices for structs

```
Foo MyStruct `json:"foo,omitzero" // optionally also `omitempty`, this is when the zero value is not valid
```

```
Foo *MyStruct `json:"foo,omitempty" // optionally also `omitzero`, this is when the zero value IS valid
```

In the WhenRequired case, this is what we should do.

When the pointer preference is always, the latter case is also used whether the zero value is valid or not

Reply:
No, because how do I expressly set the `Foo` field in this case? As a structured client? How do I make that output `{"Foo": {}}`?

Reply:
Thanks for the clarification, I was overlooking the parent struct.
Got it now. https://go.dev/play/p/Ayjn9Set966